### ISSUE TITLE
[BUG] in ICP's Kabsch solver systematically understates all reported metrics — Fix This Will Sig Imrpvoe Ur Score

### ISSUE THREAD

**Issue by ChaoyuWang04** [VAST-AI-Research/MIDI-3D#40]
````markdown
Hi, and thanks for releasing MIDI-3D — the code and the 3D-FRONT test setup were straightforward to reproduce.

While using your evaluation pipeline for an ablation study, we found what we believe is a
**rotation-transpose bug in the ICP used by `compute_metrics`**. It makes the alignment step
converge to a *worse* pose than doing nothing, which depresses every metric in Table 1.

**This should be good news:** with the fix, MIDI's own numbers improve substantially
(F-Score-S 0.46 → 0.91, CD-S 0.135 → 0.010 on the same predictions).

---

## 1. The bug

`midi/utils/metrics.py`, `compute_rigid_transform`, lines 229–231
(at `b6ec9f62532320ff0db671da84381cd60ac1db9b`):

```python
# Compute SVD
U, S, Vt = torch.svd(H)
V = Vt.transpose(1, 2)
R = torch.matmul(V, U.transpose(1, 2))
```

`torch.svd` returns **`(U, S, V)`** such that `H = U @ diag(S) @ V.mT` — it does *not*
return `Vᵀ`. (The function that returns `Vh` is `torch.linalg.svd`.) So the variable named
`Vt` already holds `V`, and line 230 turns it into `Vᵀ`.

The Kabsch solution is `R = V @ Uᵀ`, but the code computes `R = Vᵀ @ Uᵀ`.
The same `V` (i.e. `Vᵀ`) is reused in the reflection-handling branch on line 238, so that
path is affected identically.

Because `icp()` applies this rotation and accumulates it every iteration, the transform
drifts further from the optimum with each step.

---

## 2. Minimal reproduction (synthetic, no data required)

```python
import torch

torch.manual_seed(0)
A = torch.randn(1, 500, 3)
a = torch.tensor(0.7)
R_true = torch.tensor([[[torch.cos(a), 0, torch.sin(a)],
                        [0., 1., 0.],
                        [-torch.sin(a), 0, torch.cos(a)]]])
B = A @ R_true.transpose(1, 2)          # B is A rotated by a known R

AA = A - A.mean(1, keepdim=True)
BB = B - B.mean(1, keepdim=True)
H = AA.transpose(1, 2) @ BB

U, S, V = torch.svd(H)                                  # torch.svd -> (U, S, V)
R_current = V.transpose(1, 2) @ U.transpose(1, 2)       # what the code does: Vᵀ Uᵀ
R_kabsch  = V @ U.transpose(1, 2)                       # correct: V Uᵀ

for name, R in [("current (Vᵀ Uᵀ)", R_current), ("correct (V Uᵀ)", R_kabsch)]:
    print(f"{name:>18}  max|R - R_true| = {(R - R_true).abs().max():.6f}"
          f"   residual = {((AA @ R.transpose(1,2)) - BB).pow(2).sum(-1).mean().sqrt():.6f}")
```

Output:

```
   current (Vᵀ Uᵀ)  max|R - R_true| = 0.992196   residual = 2.161528
    correct (V Uᵀ)  max|R - R_true| = 0.000000   residual = 0.000000
```

---

## 3. Evidence on the actual 3D-FRONT test data

ICP's own objective is the bidirectional mean nearest-neighbour distance. We compared
three alignments of the *same* MIDI predictions against GT, over 96 test rooms:

| alignment | bidirectional mean NN distance (median) | F-Score-S |
|---|---|---|
| `icp()` as released | **0.35911** | 0.4542 |
| `icp()` with Kabsch fixed | 0.07876 | 0.9280 |
| translation only (align centroids, no rotation) | 0.07576 | 0.9035 |

**In 91 of 96 rooms, a pure translation beats `icp()` on ICP's own objective.**

We think this is decisive rather than a matter of taste: a pure translation is an element of
the rigid-transform search space that ICP optimises over, so a correctly converging ICP
cannot end up worse than one. After the fix, ICP beats translation-only in the majority of
rooms, as expected.

---

## 4. Impact on the reported metrics

Same checkpoint, same predictions, same GT, **only the alignment differs**
(random 200-room subset of the official test split, n=193 after dropping single-instance rooms):

| Metric | Paper Table 1 | released `icp()` | **Kabsch fixed** |
|---|---|---|---|
| F-Score-S ↑ | 0.502 | 0.4586 | **0.9138** |
| F-Score-O ↑ | 0.536 | 0.5397 | **0.8224** |
| CD-S ↓ | 0.080 | 0.1354 | **0.0102** |
| CD-O ↓ | 0.103 | 0.0861 | **0.0285** |

For context, our reproduction of Table 1 on the **first 100** rooms of the official test list,
using the released code unmodified, gives F-Score-S **0.5068** vs the paper's 0.502 (+0.9%)
and F-Score-O **0.5450** vs 0.536 (+1.7%) — i.e. we reproduce the paper closely, which is
what led us to believe the released evaluation path is the one behind the reported numbers.

A second, more subtle effect we observed in our own ablations: the bad alignment adds a large
**common** error floor to every arm being compared, which **compresses differences between
methods**. In one of our ablation arms an effect measured as `+5.3%, p=0.30 (n.s.)` under the
released ICP became `+14.0%, p=0.024` once the alignment was fixed. If the same holds across
methods, Table 1 may be understating the gaps between them.

---

## 5. Suggested fix

```diff
--- a/midi/utils/metrics.py
+++ b/midi/utils/metrics.py
@@
     # Compute SVD
-    U, S, Vt = torch.svd(H)
-    V = Vt.transpose(1, 2)
+    # NOTE: torch.svd returns (U, S, V) with H = U diag(S) Vᵀ — the third output is V,
+    # not Vᵀ. (torch.linalg.svd is the one that returns Vh.)
+    U, S, V = torch.svd(H)
     R = torch.matmul(V, U.transpose(1, 2))
 
     # Handle reflection case
     det_R = torch.det(R)
     ones = torch.ones_like(det_R)
-    eye = torch.eye(3, device=R.device).unsqueeze(0).repeat(R.size(0), 1, 1)
     diag = torch.diag_embed(torch.stack([ones, ones, det_R], dim=1))
     R = torch.matmul(torch.matmul(V, diag), U.transpose(1, 2))
```

(`eye` is unused; dropping it is cosmetic and optional.)

`torch.svd` is also deprecated in favour of `torch.linalg.svd`; if you migrate, note that
`torch.linalg.svd` returns `Vh`, so the correct expression there is `Vh.mT @ U.mT`.

---

## 6. What we did *not* check

To be explicit about the limits of the above:

- **We only evaluated MIDI's own released checkpoint.** We did not re-run any baseline from
  Table 1, so we have **no evidence** about whether the relative ranking of methods changes.
- Numbers above are on subsets of the official test split (n=100 and n=193), not all 1000 rooms.
- Our GT surfaces come from `load_surface`'s `.glb` branch (sampling 500k points, then
  subsampling to 20480) rather than the precomputed `.npy` files.
- `pytorch3d` was not installable on our machine (sm_120), so `knn_points` was replaced with an
  exact chunked `cdist` 1-NN. We verified it matches a brute-force reference exactly
  (identical indices and distances), and it only feeds ICP.
- Single seed (`eval_seed: 42`).

Happy to open a PR with the one-line fix, or to share the diagnostic scripts if useful.
````
